### ISSUE TITLE
Iteration problem regarding initialization with SMT and initialization problem of PSO

### ISSUE THREAD

**Issue by JieHE-2020** [figlerg/anyHR#2]
[test.zip](https://github.com/figlerg/anyHR/files/8374226/test.zip)

Hi Felix, I am now trying to use your tool to solve some SMT problems in my project. However there are two problems. You could run mytest.py in the attached test.zip file.

1. For the constraints written by me (you could simply run mytest.py and the constraints are automatically printed), I can initialize with a solution with SMT by setting 'init_point=InitPoint.SMT' in the parameters of HitAndRun class. However, then the program gets stuck when searching for a seond sample, and thus will not return a sampled result from the solution set. But if I use SMT initialization in your provided example, or change 'init_point=InitPoint.PSO' in my program, this will work.

2. The second question is not related to the attached codes. This is what I observed in my project. If I initialize a solution with PSO, for a multi-variable problem that is satisfiable but the solution set may a little hard to compute, the initialization time is quite unstable. Sometimes I need to wait for a lot of time to get the first initialized solution, but sometimes I it takes very little time. This situation often happens when the number of variables becomes large (say exceed 5). But once the initialzation solution is found, the subsequent iteration and sampling execute very quickly. I am very curious about what happens in the initialization phase of PSO, because its speed now becomes a bottleneck of my project.

I will be very happy if you could help me with these problems. Thank you very much!

**Comment by figlerg** [comment]
Hi, thanks for the message!

1. Your constraints appear to be fine. The way hit-and-run works sometimes takes a very long time, if the constraints define a very 'narrow' space. A way around this would be to use the 'shrinking' optimization option at creation of the HitAndRun object, as seen in the following snippet where i modified the end of your code:
```python 
# build hr object

from anyHR.hit_and_run.hit_and_run import Shrinking

y_hr = HitAndRun(c_y, y_bounds, shrinking = Shrinking.SHRINKING)



# generate samples
# ps: the below could also be achieved by calling the wrapper 'y_hr.sampler(100,10)'

samples = []
nb_samples = 100
mixing = 10
for j in range(nb_samples * mixing):
    sample, rejections = y_hr.next_sample()

    # do some mixing in between samples
    if j % mixing == 0:
        samples.append(sample)

y_coordinate_list = random.sample(samples, 1)[0]
print(y_coordinate_list)
```

For me it now produces samples reasonably fast.

2. This is less easy for me to adress: The particle swarm algorithm we use here is merely an optimization algorithm, where the hypercube is searched for any point that is sat. Here, we use pyswarm with some fixed parameters.
If the sampled set is very small compared to the hyperrectangle (which easily happens in many dimensions), the PSO may not find a solution. One could play around with the values in pso/pso.py or try other ways of finding a sat point.

In any case, I should perhaps improve the pso interface and error handling.



Let me know if any of this helped.

**Comment by JieHE-2020** [comment]
[new_test.zip](https://github.com/figlerg/anyHR/files/8396407/new_test.zip)

Hi Felix, many thanks for your explanation and help!

I tried to use 'shrinking' optimization for my minimal testing case, ant it now works when I initialize a solution with SMT.

Then I tried to test whether this method could work for more complex constraints. Please see 'mytest2.py' in new_test.zip.

This time there are 8 variables, and the solution space is more narrow. However, 'shrinking' optimization unfortunately does not make the SMT initializer successfully sample the next solutions, even though I changed the direction of sampling to CDHR. Problem 2 regarding the PSO initializer still exists, since 'shrinking' does not affect its initialization phase.

Thank you in advance.

**Comment by figlerg** [comment]
Hi, sorry for the wait!

Thanks for bringing this to my attention, this was a bug (or a badly done user interface) where the SMT solver would not accurately enforce the bounding box. I released a new version where this is fixed, let me know if the new version works for your examples! (I tested it on the example you sent me and it worked.)

Yes, PSO and shrinking are not interfering at all. In my experience, you should use SMT for all applicable situations (anything that is linear, for example). PSO does not guarantee to find a point within the given budget of iteration steps, and most likely will not if the set is very small. It also will not differentiate between a small set and an empty set (impossible constraints), whereas SMT will tell you if your constraints are not satisfiable.

Best,
Felix

**Comment by JieHE-2020** [comment]
Hi Felix, thank you very much for the fix. I tried the new version and it now works perfectly in my example!
Yes you are right, since my constraint is all linear, I can use SMT all the way around for initialization.

For the PSO initializer, I also tested it. Sometimes I can get the starting point and successfully continue subsequent iteration and sampling, but in most cases there will be an error in the start point searching phase as follows:

Traceback (most recent call last):
  File "/Users/jie.he/Desktop/programs/test copy/mytest2.py", line 106, in <module>
    x_hr = HitAndRun(c_x, x_bounds, shrinking=Shrinking.SHRINKING, init_point=InitPoint.PSO)
  File "/Users/jie.he/.local/lib/python3.9/site-packages/anyHR/hit_and_run/hit_and_run.py", line 63, in __init__
    self.starting_point = self._starting_point_pso()
  File "/Users/jie.he/.local/lib/python3.9/site-packages/anyHR/hit_and_run/hit_and_run.py", line 354, in _starting_point_pso
    raise Exception('Could not find a starting point in the set.')
Exception: Could not find a starting point in the set.

This is consistent with your explanation, maybe it is due to the small solution set. But anyway it would not affect my use of this library because I can use SMT alternatively. Thank you very much!

Best,
Jie
